gr-digital/lib/vblast_decoder_cc_impl.cc

```diff
@@ -0,0 +1,237 @@
+      }
+    }
+
+    void
+    vblast_decoder_cc_impl::equalize_symbol(gr_vector_const_void_star input,
+                                            gr_complex* out,
+                                            uint32_t offset,
+                                            uint32_t length) {
+      std::fill(out, &out[length*d_num_inputs], 0.0);
+      for (int n = 0; n < d_num_inputs; ++n) {
+        gr_complex *in = &((gr_complex *) input[n])[offset/d_num_inputs];
+        for (unsigned int i = 0; i < length; ++i) {
+          for (int j = 0; j < d_num_inputs; ++j) {
+            out[i*d_num_inputs + j] += d_mimo_equalizer[j][n] * in[i];
+          }
+        }
+      }
+    }
+
+    int
+    vblast_decoder_cc_impl::work(int noutput_items,
+        gr_vector_const_void_star &input_items,
+        gr_vector_void_star &output_items)
+    {
+      gr_complex *out = (gr_complex *) output_items[0];
+      uint16_t nprocessed = 0; // Number of read and written items.
+
+      // Collect all tags of the input buffer in the vector 'tags'.
+      get_tags_in_window(tags, 0, 0, noutput_items);
+
+      uint16_t symbol_length; // Number of items in the current symbol.
+
+      if(tags.size() == 0){ // Input buffer includes no tags at all.
+        // Handle all samples in buffer as they belong to the current symbol.
+        symbol_length = noutput_items;
+        equalize_symbol(input_items, out, nprocessed, symbol_length);
+        nprocessed += symbol_length;
+      } else { // Input buffer includes tags.
+        if (tags[0].offset - nitems_read(0) > 0){
+          /* There are items in the input buffer, before the first tag arrives,
+           * which belong to the previous symbol. */
+          symbol_length = (tags[0].offset - nitems_read(0))*d_num_inputs;
+          equalize_symbol(input_items, out, nprocessed, symbol_length);
+          nprocessed += symbol_length;
+        }
+        // Iterate over tags in buffer.
+        for (unsigned int i = 0; i < tags.size(); ++i) {
+          // Calculate the number of items before the next tag.
+          if (i < tags.size() - 1) {
+            // This is not the last tag.
+            symbol_length = (tags[i + 1].offset - tags[i].offset)*d_num_inputs;
+          } else {
+            // This is the last tag.
+            symbol_length = noutput_items - (tags[i].offset - nitems_read(0))*d_num_inputs;
+          }
+          // Check the key of the tag.
+          if (pmt::symbol_to_string(tags[i].key).compare("csi") == 0) {
+            // Get CSI from 'csi' tag.
+            for (unsigned j = 0; j < pmt::length(tags[i].value); ++j) {
+              d_csi[j] = pmt::c32vector_elements(pmt::vector_ref(tags[i].value, j));
+            }
+            // Calculate the weighting vector for the next symbol with the received CSI.
+            update_mimo_equalizer();
+          } else if (pmt::symbol_to_string(tags[i].key).compare("snr") == 0 && d_equalizer_type.compare("MMSE") == 0) {
```

Comment:
same as above.

Reply:
The problem with comparing the string really is that PMT is horrific. The `symbol_to_string` function does a dynamic cast of the `pmt_base` to a `pmt_symbol`. That's an awful lot of indirection just to compare. Also, in a lot of cases (I think this might be one), this requires a copy of the `std::string` contained in that PMT to be made (did I ever mention how bad an idea that design is? Like, have "polymorphic type", but it not be properly polymorphic, and it also just inefficiently wrap std types, for communication within the same process?).

`std::string` copies are relatively expensive and you try to avoid them in the signal processing part, because they require allocation and deallocation of memory.